gems/sorbet-runtime/lib/types/private/methods/_methods.rb

```diff
@@ -345,12 +352,34 @@ def self.run_sig(hook_mod, method_name, original_method, declaration_block)
     signature
   end
 
-  def self.run_builder(declaration_block)
-    builder = DeclBuilder.new(declaration_block.mod, declaration_block.raw)
-    builder
-      .instance_exec(&declaration_block.blk)
-      .finalize!
-      .decl
+  def self.run_builder(declaration_block, method_name)
+    return if processing_method?(declaration_block.mod, method_name)
+
+    builder = register_method_being_processed(declaration_block, method_name)
+    begin
+      builder.run!(&declaration_block.blk)
+    ensure
+      unregister_method_being_processed(declaration_block.mod, method_name)
+    end
+
+    builder.decl
+  end
+
+  def self.processing_method?(mod, method_name)
+    key = method_owner_and_name_to_key(mod, method_name)
+    @methods_being_processed[key]
+  end
+
+  def self.register_method_being_processed(declaration_block, method_name)
+    key = method_owner_and_name_to_key(declaration_block.mod, method_name)
+    @methods_being_processed[key] = true
+
+    declaration_block.decl_builder
+  end
+
+  def self.unregister_method_being_processed(mod, method_name)
+    key = method_owner_and_name_to_key(mod, method_name)
+    @methods_being_processed.delete(key)
```

Comment:
It's not exactly cheap to continually compute `method_owner_and_name_to_key`, and we're doing it three times in `run_builder`.  Can we at least compute the key once, and pass it in to each of these functions?

But why are we computing such a complex key?  Is the module itself not enough?  Or is the concern that we might wind up doing signature building for the same module, but in different files and different methods, and we really might need to do both of them simultaneously?  Is there a test for that?

Reply:
Let me know if you want this test added to the PR.This is the diff against the `test` commit on this branch:

```
gems/sorbet-runtime/test/types/fixtures/duplicate_sig_eval/define_method_with
_sig.rb
@@ -4,5 +4,8 @@ class DefineMethodWithSig

     sig {params(arg: T.nilable(T.class_of(Opus::Types::Test::DuplicateSigEvalSandb
ox::CallsMethodUponLoading))).void}
     def self.duplex(arg); end
+
+    sig { raise "sig is evaluated" }
+    def self.foo; end
   end
 end

gems/sorbet-runtime/test/types/fixtures/duplicate_sig_eval/calls_method_upon_
loading.rb
@@ -1,5 +1,5 @@
 module Opus::Types::Test::DuplicateSigEvalSandbox
   class CallsMethodUponLoading
-    Opus::Types::Test::DuplicateSigEvalSandbox::DefineMethodWithSig.duplex(self)
+    Opus::Types::Test::DuplicateSigEvalSandbox::DefineMethodWithSig.foo
   end
 end
```

As before we'd trigger `duplex` from the test runner. Which would load the sig for `duplex` which in turn loads the `CallsMethodUponLoading`. However, now that class calls the `foo` method instead of `duplex`.

Without a method specific key the current implementation would not evaluate the sig for `foo` and not raise.

Reply:
Can we have both the test that has `duplex` and the test that has `foo`?  I don't see why it has to be one or the other.

Reply:
Added the test. Used `check_level` instead of `raise` to ensure signature was fully evaluated.

I can take a closer look at the stacktrace you provided however, upon first glance it seems consistent with the initial reproduction steps by Jake. I did test this implementation with `T::Utils.run_all_sig_blocks` before and it was working 🤔 .